[2단계 - 장바구니 협업] 마르코(장원석) 미션 제출합니다.

wonsss opened the pull request:
안녕하세요. 발리스타님 😄 백엔드와 협업을 하며 재밌고 바쁜 한 주였네요.
함께 작업한 코드가 아닌 이전 미션에서 페어가 작성한 코드를 수정하려고 하니, 생각했던 것보다 어려웠던 것 같아요. 그러나 다른 분의 코드를 이해하려고 노력하면서 많은 공부가 되어 좋았습니다. 그리고 백엔드와 협업하며 http 메서드, 상태코드를 공부하고 mixed-content와 cors 이슈 등에 부딪히면서 많은 것을 배웠습니다. 

[참고 : 페어 및 백엔드와 협업하며 작성한 API 명세](https://polyester-stay-438.notion.site/Shopping-Cart-API-80ff27ed17b14608ac58d1edd880744c)

바쁘실텐데 리뷰에 힘써주셔서 감사합니다. 잘 부탁드립니다.

## 데모페이지
### [마르코 데모페이지](http://react-shopping-cart-heroku.herokuapp.com)
- 백엔드 크루의 API를 연결하였습니다.

##  로컬에서는 MSW
- 로컬에서 npm install 및 npm run start 하면 MSW 구동이 가능합니다.

## 구현 기능
- [x] 사용자 인증 관련 API
    - [x] 1. 회원 가입(POST)
    - [x] 2. 회원 탈퇴(DELETE)
    - [x] 3. 닉네임 수정(PATCH)
    - [x] 4. 비밀번호 수정(PATCH)
    - [x] 5. 로그인(POST)
    - [x] 6. 회원 조회(GET)
- [x] 장바구니 관련 API
    - [x] 1. 상품 목록 가져오기(GET)
    - [x] 2. 특정 상품 가져오기(GET)
    - [x] 3. 장바구니 목록 가져오기(GET)
    - [x] 4. 장바구니 목록에 상품 추가 및 수량 수정하기(PUT)
    - [x] 5. 장바구니 내 선택된 상품 삭제(DELETE)
    - [x] 6. 주문 추가하기(POST)
- [x] 사용자 인증 API에 대하여 msw로 모킹하여 로컬에서 테스트할 수 있다.
- [x] 백엔드 크루가 제공한 사용자 인증 API로 호스트를 변경하여 테스트할 수 있다.
- [x] 스낵바 구현
- [x] 새로고침 시 로그인 유지(GET 요청하여 accessToken 검증)
- [x] 인가되지 않은 유저의 특정 페이지 및 행동 접근 막기('/account', 장바구니 추가하기)
- [x]  로그인된 유저의 특정 페이지 접근 시 리다이렉션('/login', '/signup') - 이미 로그인됐으므로 불필요한 접근
- [x] 각 인풋 아래에 실시간 유효성 검사 메시지 표시

## 추가 리팩토링한 점
- [x] 리덕스 덕스 패턴 적용
- [x] (API, 리덕스 등에 대한) 컨트롤러스러운 커스텀 훅을 만들어 분리하여, 페이지 파일에서 직접 API 처리하지 않기

## 고민
- 기존에 페이지 파일에 있던 API 관련 로직을 커스텀훅(use***API)으로 빼주었습니다. 이렇게 분리한 커스텀훅들 중에서 단 한 번만 사용되며 특정 페이지 기능과 밀접히 연관되어 있다고 판단되는 훅들의 경우, 해당 페이지 폴더 밑으로 이동시켰습니다. 가독성을 고려하여 이런 방식으로 해봤는데, 리뷰어님의 의견은 어떤지 궁금합니다.
- 그리고 해당 API 관련 커스텀훅 안에 `스낵바` 로직도 함께 두었습니다. 그런데 어차피 해당 커스텀훅이 API 통신 결과의 `error`를 반환하므로, `스낵바` 로직을 커스텀훅에서 빼서 관련 `페이지` 파일에서 직접 사용하는 것은 어떻게 생각하시나요? 스낵바 뿐만 아니라 라우팅을 하는 내비게이트도 마찬가지입니다. 이렇게 고민하는 이유는 스낵바나 내비게이트 같은 로직은 `API, 리덕스 등` 데이터베이스 관련 컨트롤러 역할의 커스텀훅에 어울리지 않을 것 같기 때문입니다. 
- API url을 `env`로 관리하고 gitignore했다가, 어차피 해당 API url이 개발자도구 네트워크 탭 등에 다 노출될텐데 굳이 숨길 필요가 있나 싶은 생각이 들어서 API url을 env로 관리하며 숨기지 않고 코드 내에 직접 넣었습니다. API key 같은 것들은 env로 관리하고 숨길 필요가 있겠으나, 이 프로젝트에서 같은 API url은 숨기려는 노력이 필요할까요..?

wonsss committed:
chore: step 2준비

wonsss committed:
fix: 새로고침 시 로그인 풀리던 문제 해결

리덕스에 isLoading 상태 추가

wonsss committed:
chore: 콘솔로그 삭제

wonsss committed:
build: 환경변수에 사용자인증 api url 추가

wonsss committed:
chore: 배포 오류 수정 시도

wonsss committed:
fix: axios 인스턴스 및 env 개발과 프로덕션 모드 분기

wonsss committed:
fix : Mixed content 에러 해결

에러메시지 - This request has been blocked; the content must be served over HTTPS.

wonsss committed:
fix: 잘못된 상수 수정

wonsss committed:
fix:  쿠키 cors 통신 설정

wonsss committed:
refactor: 로컬 development 환경에서 msw 적용되도록 설정

wonsss committed:
fix: netlify 배포 시 리다이렉션 설정

wonsss committed:
feat:   'Access-Control-Allow-Origin' 설정

wonsss committed:
fix : Mixed content 에러 관련 upgrade-insecure-requests

에러메시지 - This request has been blocked; the content must be served over HTTPS.

wonsss committed:
feat: redux 미들웨어(composeWithDevTools, logger) 설치

wonsss committed:
fix: 결제예상금액 오류 수정

wonsss committed:
feat:리덕스에 orderList 상태 추가 및 orderPage 라우팅

wonsss committed:
feat: [todo] 상품 API 추가할 예비 코드 주석 추가

wonsss committed:
fix: cors 에러를 withCredential 설정으로 해결

wonsss committed:
docs: step2 할 일 정리

wonsss committed:
refactor: axios 인스턴스에 헤더 추가하고 중복 제거

wonsss committed:
feat: 세션 쿠키로 변경

wonsss committed:
refactor: 에러메시지 수정

wonsss committed:
fix: 회원정보 수정 페이지에 로그인되지 않은 유저가 접근할 경우 막기

wonsss committed:
feat: 인풋 auto focus

wonsss committed:
feat: 회원가입 후 리다이렉션되는 로그인 페이지의 이메일 인풋에 이메일 넣기

wonsss committed:
refactor: apiClient위치를 apis 폴더로 이동

wonsss committed:
feat: thunk 설치 및 createPromiseThunk 유틸 작성

wonsss committed:
feat: useReduxState 커스텀훅 작성

wonsss committed:
refactor: useDispatch 사용

wonsss committed:
refactor: 폴더이름과 중복되는 파일명을 index로 변경

wonsss committed:
refactor: import 문 정리

wonsss committed:
refactor: 컴포넌트 import 정리

wonsss committed:
refactor: import문 순서 정리

라이브러리(커스텀훅) > 컴포넌트, 스타일 > 유틸

wonsss committed:
refactor: 환경에 따라 부르는 API 로직 단순화 및 백엔드 크루 API추가

wonsss committed:
feat: 커스텀 에러 추가 및 에러 분기 다양화

서버 API에서 보내주는 에러코드와 에러메시지를 받아 분기하여 표시

wonsss committed:
fix: 쿠키에 저장된 accessToken에 대한 decode 해결

wonsss committed:
chore: 불필요한 코드 삭제

wonsss committed:
feat: msw에서 입력한 이메일이 db 존재 여부 확인

wonsss committed:
feat: API 명세 변경에 따른 URI 수정(닉네임, 비밀번호 변경 관련)

wonsss committed:
docs: API 투두리스트 수정

wonsss committed:
feat: [API] 1. 상품 목록 가져오기(get)

wonsss committed:
feat: 닉네임 수정 및 비밀번호 수정 API 모킹 변경

wonsss committed:
feat : [API] 장바구니 목록 가져오기 및 추가와 수정 msw 구현

wonsss committed:
feat: 장바구니 결제예상금액 박스를 sticky 포지션으로 변경

wonsss committed:
feat: 수량 조절 버튼을 drag 할 수 없도록 css 수정

wonsss committed:
feat: 장바구니 내 선택된 상품 삭제(MSW API)

wonsss committed:
feat: API 수정

wonsss committed:
fix: id를 productId로 변경

wonsss committed:
build: 빌드 scripts 변경

wonsss committed:
build: 빌드 scripts 변경

wonsss committed:
refactor: 불필요한 useLogout 사용 코드 제거

wonsss committed:
feat: 주문 추가하기(POST API)

wonsss committed:
chore: 불필요한 코드 삭제

wonsss committed:
refactor: Redux의 Ducks 패턴 적용

wonsss committed:
fix: api 명세에 따라 productId를 id로 변경

wonsss committed:
feat: msw에서 장바구니 주문 추가 모킹

wonsss committed:
refactor: useGetProductsAPI 커스텀훅 분리

wonsss committed:
refactor: useGetCartAPI 커스텀훅 분리

wonsss committed:
refactor: 커스텀훅에 useCallback 적용

wonsss committed:
refactor: useGetProductAPI 커스텀훅 분리

wonsss committed:
refactor: useCartAPI 커스텀훅 분리

wonsss committed:
refactor: useDeleteProduct(s)API 커스텀훅 분리

wonsss committed:
refactor: useOrderAPI 커스텀훅 분리

wonsss committed:
refactor: useTotalPrice 커스텀훅 분리

wonsss committed:
refactor: API 관련 커스텀훅 폴더링

wonsss committed:
refactor: useDeleteAccountAPI 커스텀훅 분리

wonsss committed:
chore: 불필요한 import 삭제

wonsss committed:
react-create-app on Heroku

wonsss committed:
heroku

wonsss committed:
heroku

wonsss committed:
heroku

wonsss committed:
heroku

wonsss committed:
heroku push 서버 선택 페이지'

wonsss committed:
fix: 로그인 직후 로그아웃되지 않던 문제 해결

wonsss committed:
chore: 불필요한 delete post api 삭제

wonsss committed:
chore: env 환경에 따라 모킹 여부 설정

wonsss committed:
refactor: routes 상수화

wonsss committed:
feat: signup 페이지 커스텀훅 분리

wonsss committed:
chore: 잘못된 폴더명 정정

wonsss committed:
feat: 백엔드 크루 api 선택 후 새로고침 용이하게 변경

wonsss committed:
refactor: AUTH_API_URL를 constants 파일로 이동

wonsss committed:
feat: login 페이지 커스텀훅 분리

wonsss committed:
chore: 데모페이지 주소 추가 및 순서 정리

wonsss committed:
refactor: 불필요한 코드 정리 및 이동

wonsss committed:
refactor: passwordEdit 페이지 커스텀훅 분리

wonsss committed:
refactor: 서버 선택하는 페이지 라우트 위치 변경

wonsss committed:
refactor: AccoutPage 커스텀훅 분리

wonsss committed:
heroku update

wonsss committed:
chore: package 및 사용하지 않는 파일 정리

wonsss committed:
fix: 로그인 후 액세스토큰이 요청 헤더에 담기도록 수정

wonsss committed:
fix: 로그인 후 액세스토큰이 요청 헤더에 담기도록 수정

wonsss committed:
chore: 기본 API 지정

woowapark commented on .storybook/preview.js:
```diff
@@ -1,11 +1,11 @@
 import store from 'store/store';
-import { doInitializeProductList } from 'actions/actionCreator';
 import { dummyProductList } from 'dummy_data';
+import { doInitializeProductList } from 'reducers/cart.reducer';
```
`initialize-`로도 충분할 것 같은데 `do-`가 붙은 이유가 있을까요?-?

woowapark commented on REQUIREMENTS.md:
```diff
@@ -18,10 +77,10 @@
   - 닉네임
     - 2자리 이상 10자리 이하이어야 한다.
-- [ ] 회원탈퇴
+- [x] 회원탈퇴
```
와우 REQUIREMENTS를 엄청 꼼꼼히 작성해주셨네요 👍

wonsss replied:
감사합니다! :)

woowapark commented on package.json:
```diff
@@ -23,15 +13,14 @@
     "test": "react-scripts test",
     "eject": "react-scripts eject",
-    "deploy": "gh-pages -d build",
+    "predeploy": "npm run build",
```
위에 있는 build와 같은 명령어인 것 아닌가용?

wonsss replied:
배포를 이전 페어 방식인 gh-pages로 하지 않고 heroku 등 다른 방식을 이용할 생각이어서, deploy 명령어를 삭제했어요. 그런데 npm run deploy 명령어를 입력했을 때 deploy가 수행되기 전에 build를 하는 `predeploy`는 깜박하고 그대로 뒀었네요.  predeploy 명령어도 현재 필요 없으니 삭제했어요! 😄 
[chore: 불필요한 명령어 삭제](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/9789f831865e3b27abe6c648312f5b32700d4f7e)

woowapark commented on src/App.js:
```diff
@@ -1,43 +1,40 @@
+import { ROUTES } from 'utils/constants';
+import { getCookie } from 'utils/cookie';
+import apiClient from 'apis/apiClient';
+import { doLogin } from 'reducers/auth.reducer';
```
오옹 🤔 
- style이나 test 외에 요런식의 파일 네이밍은 잘 보지 못했던 것 같아 약간 어색하게 느껴지는데 마르코는 어떠신가요?-? + rootReducer같은 경우는 카멜케이스로 네이밍되어 있어서 컨벤션을 한쪽으로 맞추면 좋을 것 같아요~! 🙂 
- 여기도 굳이 'do-'는 없어도 되지 않을까요? '로그인한다한다'같은 느낌이 아닐지..!

wonsss replied:
- 말씀해주신 것처럼 컨벤션이 다른 네이밍 방식이 어색하네요. `auth.reducer` 와 같은 파일 네이밍을 rootReducer와 동일한 카멜케이스 네이밍 방식으로 변경했어요! 
[refactor: reducer 파일을 카멜케이스 네이밍 방식으로 변경](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/5ed6853ae31bc84c8635cadf3cbdddd454ca3d42)
- 그리고 저도 action creator의 네이밍 컨벤션이 `do-`로 시작하는 게 어색하다는 점 공감합니다! 불필요하게 do 접두사를 붙여준 것 같아요. 이것 또한 수정했어요!
[refactor: action creator 네이밍 컨벤션 변경('do' prefix 삭제)](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/7ec7a2dac61001b571fffd4357dee43daf75c9ba)

woowapark commented on src/App.js:
```diff
@@ -1,43 +1,40 @@
-        },
-      });
-      store.dispatch(doLogin({ nickname: response.data.nickname }));
+      const response = await apiClient.get('/customers');
```
apiClient를 분리해서 컴포넌트에서 비동기 로직을 한번 분리해주셨네요 🙂👍      
이렇게 분리해주셨다면, 위에서`accessToken` 유무에 따른 return 처리도 apiClient의 책임으로 볼 수 있지 않을까요?      
- 컴포넌트가 어떤 api를 호출할 때 accessToken이 필요하고, 어떤 api는 필요하지 않은지에 대해 신경쓰지 않아도 되는 게 좋을 것 같아요 
  - api를 호출하는 로직을 수정해야 할 때, api 쪽만 보면 될지 컴포넌트 로직까지를 함께 신경써야 할지를 고민해보시면 좋을 것 같습니다 🙂

wonsss replied:
accessToken 유무에 따른 return 처리도 apiClient의 책임으로 보는 게 맞을 것 같아요!
그런데 현재 작성한 apiClient가 객체여서 accessToken 유무에 따라 return 처리를 스스로 어떻게 하게 할 수 있을지 고민되네요. apiClient를 함수로 만들어야 할까요..? 😮

woowapark replied:
요건 아래 다른 피드백에서 class로 바꾸면서 무효해진 질문이시겠죵?ㅋㅋ 
인스턴스를 여러 개 만들 게 아니고 그냥 함수들의 묶음 용도라면 단순 객체로만 만들어도 문제 없을 것 같고요 :) 
```js 
const apiClient = axios.create();

export default {
  fetchUser() {
    return apiClient.get('/customers')
  }
}
```

- 1차로 드린 피드백과 유사하지만 api 레이어를 분리하셨다면, 어떤 엔드포인트로 요청을 보내는지 / 해당 API에 필요한 권한이 무엇인지 등을 모두 api 레이어에서 처리하도록 하는 게 어떨까요? 🙂          
컴포넌트에서는 '구체적인 엔드포인트, 권한 정보는 모르지만 사용자 정보를 불러오는 api를 호출해야해'만 알면 되고, 실제 api 호출에 필요한 정보는 apiClient 내로 감추는 게 이후 유지보수에도 좋지 않을까 싶어요 :)       
사용자 정보를 불러오는 api가 뭔가 변경이 좀 되더라도(이건 UI의 변경이 아니니), apiClient쪽에서만 한번에 수정하면 되고 컴포넌트에서는 신경쓰지 않아도 되도록이요!
요건 머지 후에 따로 고민해보셔도 좋을 것 같습니당! 
- axios [interceptor](https://axios-http.com/docs/interceptors)를 참고해보셔도 좋구요


+) React로 넘어오면서 모든 로직을 React에서 사용되기 위한 형태(컴포넌트, 훅)로만 생각하고 유틸성인 것들만 일반 js/ts로 생각하게 되기 쉬운 것 같은데, 렌더링 및 React의 라이프사이클과 상관없는 부분은 일반 js/ts로도 생각해보면 좋을 것 같습니당~!

woowapark commented on src/App.js:
```diff
@@ -1,43 +1,40 @@
-      deleteCookie('accessToken');
-      store.dispatch(doLogout());
-      renderSnackbar(MESSAGE.NO_AUTHORIZATION, 'FAILED');
+      console.error(error);
```
사용자는 어떤 일이 일어난 건지, 이 서비스를 다시 사용하려면 어떤 동작을 하면 될지 알 수 없지 않을까요?-?

wonsss replied:
App이 마운트될 때만 한 번(새로고침 등) 실행되는 getAccount() 함수는 사용자가 아직 쿠키에 갖고 있는 accessToken을 검증하여 자동로그인을 도와주는 목적을 지녔어요. 만약 accessToken이 유효하지 않다면 자동로그인이 되지 않은 상태가 유지돼요. 일반적으로 여러 사이트에 접속했을 때, 로그인 상태가 풀린 경우라면 스낵바 등 어떤 알림이 나오진 않고 로그인이 필요한 버튼을 누르면 로그인 페이지가 표시되는 UI를 경험하여, 이렇게 구현했어요. 제 의도는 이러했는데, 공원은 어떻게 생각하시나요? 😄

woowapark commented on src/App.js:
```diff
@@ -47,16 +44,18 @@ function App() {
+            <Route path={ROUTES.SIGNUP} element={<SignupPage />} />
+            <Route path={ROUTES.ACCOUNT} element={<AccountPage />} />
           </Route>
+          <Route path={ROUTES.SERVER} element={<SeverSelectPage />} />
```
SeverSelectPage가 뭔가 했더니 `Server-`의 오타였군요ㅋㅋ 👀

wonsss replied:
ㅎㅎㅎㅎㅎ앗 이 오타 고쳤다고 생각했는데, 아직도 남아있었네요! 확인 감사합니다. 수정했어요 😄

woowapark commented on src/components/CartProductItem/index.jsx:
```diff
@@ -1,26 +1,18 @@
-import useProduct from 'hooks/useProduct';
+// @ts-nocheck
```
TS를 적용한 곳과 적용하지 않은 곳을 나누는 기준이 있을까요?

wonsss replied:
미션 시작할 때 페어와 야심차게 "앞으로 작성하는 코드는 TS를 적용해보자!"라고 결의를 다졌는데, 미션 기간이 촉박하다보니 TS는 뒷전이 되고, 일단 구현하는데 급급해져서 신경쓰지 못했습니다.
즉, ㅠㅠ기준은 없습니다..흑흑 슬프네여

woowapark replied:
ㅋㅋㅋ프로젝트에서는 한번 다시 결의를 다져보시지요~! 💪

woowapark commented on src/components/CartProductItem/index.jsx:
```diff
@@ -31,13 +23,15 @@ const CartProductItem = ({ id, quantity }) => {
             if (quantity > 1) {
-              store.dispatch(doPutProductToCart({ id, quantity: quantity - 1 }));
+              putCart(productId, quantity - 1);
             }
```
UI에서 너무 많은 상세사항을 알고 있는게 아닐까 싶은데 어떠신가요? 🙂      
훅으로 분리까지 해주었다면, UI에서는 그냥 신경쓰지 않고 `decreaseQuantity()` 같은 함수를 호출만 하고,      
- `if (quantity > 1)` 나  
- `quantity - 1`     
같은 부분은 해당 함수에서 알아서 하도록 하면 어떨까요?

wonsss replied:
오오 좋네요! 커스텀훅으로 분리를 했으니, UI에서 너무 많은 상세사항까지 책임지도록 하지 말고 커스텀훅에게 전가하는 것이 더 깔끔한 것 같아요. 수정했어요!
[refactor: usePutCartAPI커스텀 훅에 UI의 기능 옮기기](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/a220078bc673a3ea99d4d1726e67c7fbc676bdaf)

woowapark commented on src/utils/constants.js:
```diff
@@ -36,8 +33,42 @@ const MESSAGE = {
+  이프: 'http://ec2-3-39-234-109.ap-northeast-2.compute.amazonaws.com:8080',
+  찬: 'http://ec2-3-34-130-116.ap-northeast-2.compute.amazonaws.com:8080',
+  더즈: 'http://15.164.211.129:8080',
+  토르: 'http://ec2-15-164-232-166.ap-northeast-2.compute.amazonaws.com:8080',
```
### 질문 3 env 관련 답변  
- 배포된 리얼 서버의 url이라면 마르코가 적어주신 것처럼 '숨겨야'하는 정보는 아닙니다. 사용자가 접근하는 페이지에서 네트워크 탭을 열면 이미 다 노출되긴 하니까요 🙂 
- 다만 외부에 공개되면 안되는 개발 서버의 주소이거나, 특정 ip 를 담고 있다면 보안상의 문제가 생길 수 있을 것 같네요. 누군가 악의적인 요청을 보낼 수도 있고, 개발서버이기 때문에 보안 처리를 엄격히 안하는 부분도 있을텐데 예상치 못한 공격에 뚫릴 수도 있고요. 
- 덧붙여, api도 local, dev, production 등 환경에 따라 다른 base url을 사용해야 하는 경우가 있으니 그런 환경에 따른 구분을 위해서 사용하는 측면도 있습니다.

wonsss replied:
`배포된 리얼 서버 주소`이냐 `외부에 공개되면 안되는 개발 서버 주소` 이냐에 따라, 숨겨야 하는 정보인지 판단이 달라지는 거군요.
그리고 숨길 필요가 없는 배포된 리얼서버 주소라고 하더라도 환경에 따라 다른 base url을 자동으로 주입시켜주기 위해 env를 사용하는 측면도 있군요! 명쾌한 설명 감사합니다. 👍

woowapark commented on src/page/LoginPage/index.jsx:
```diff
@@ -0,0 +1,42 @@
+          setInputValue={setPassword}
+        />
+        <AuthButton actionType="Login" action={login} isDisabled={!isFulfilled} />
+        <GuideText guide="Don’t have an account?" destination="Sign up" path="/signup" />
```
폼 제목은 한국어인데 폼 라벨이랑 여기는 또 영어라서 당황했는데요ㅋㅋ 
언어가 다른 건 의도하신 걸까요?      
가능하면 언어는 하나로 맞추는 것이 자연스러울 것 같네요!ㅋㅋ

wonsss replied:
ㅎㅎㅎ 페어와 간지나는 사용자인증 페이지를 만들어보려고 영어를 섞어봤는데, 다시 보니 말씀대로 자연스럽지는 않네요! 앞으로 언어는 하나로 통일하여 구현하도록 하겠습니다! 감사해요

woowapark commented on src/hooks/useCheckAuth.js:
```diff
@@ -0,0 +1,21 @@
+  }, [isAuthenticated, isFulfilled, isLoading, navigate, renderSnackbar]);
+};
+
+export default useCheckAuth;
```
아마 인증이 필요한 페이지마다 요 훅을 호출해주어야 할텐데요,      
인증된 경우에만 컴포넌트를 렌더하도록 하는 private route(혹은 protected route, authenticated route 등등으로 검색해보셔도 예시를 많이 보실 수 있을 거에요)를 만들어서 적용해보는 건 어떨까요? 
```jsx
// 힌트: 아래와 같이 사용했을 때 PrivateRoute 는 인증된 상태에서만 접근되고 + HOME으로 리다이렉트될 수 있도록이요 :) 
<Route path={ROUTES.HOME} element={<ProductListPage />} />
<PrivateRoute path={ROUTES.CART} element={<CartPage />} />
``` 
요런식으로 라우팅 설정을 할 때 아예 페이지 단위의 접근 권한을 관리할 수 있도록이요 🙂 
- 현재 페이지가 인증된 유저만 접근할 수 있는지 여부는 컴포넌트의 관심사가 아닐 수 있고, 
- 접근 권한이 없다면 아예 컴포넌트 렌더 로직에 진입하기 전에 막히는 것이 맞을 것 같아요

wonsss replied:
라우팅 설정 시 페이지 단위의 접근권한을 관리할 수 있도록 PrivateRoute와 PublicRoute를 만들어 적용해봤습니다! 말씀대로, 접근 권한이 없다면 컴포넌트 렌더 로직에 진입하기 전에 다른 곳으로 라우팅해주는 게 더 나은 사용자 경험인 것 같아요! 감사합니다. 😄 
[feat: PrivateRoute, PublicRoute 적용](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/4eb5aff60a359a58951d1546e0d93a297b7490e8)

woowapark replied:
Route 하위에서 별도의 컴포넌트를 하나 더 두지 않고,      
PrivateRoute가 (이름처럼) 아예 `Route`를 반환해주어도 좋을 것 같네용~ 🙂 
```jsx
return isAuthenticated ? <Route {...props} /> : <Navigate to={PATH.LOGIN} />;
```

woowapark commented on src/components/CartProductItem/useDeleteProductAPI.js:
```diff
@@ -0,0 +1,34 @@
+    try {
+      await apiClient.delete('/cart', { data: { productIds: [productId] } });
+      dispatch(doDeleteProductFromCart({ productId: productId }));
+      renderSnackbar(MESSAGE.REMOVE_CART_SUCCESS, 'SUCCESS');
```
## 질문 2 답변 
커스텀 훅에서 어떤 것을 제공하고 싶은 의도였는지에 따라 다를 순 있을 것 같아요. 
만약 이런저런거 아무것도 신경쓰지 않고 모든 걸 알아서 해주는 훅으로 만들고 싶었다면 지금처럼 스낵바까지도 알아서 띄우는 게 맞을 것이고요. 
다만 개인적으로는 저도 이런 경우에는 snackbar와 같은 UI 로직은 커스텀 훅으로 묶지 않는 편이 좋다고 생각합니다. 
- '어떤 컴포넌트에서 이 훅을 사용할 때는 스낵바를 띄우지 않고 다른 처리를 하고 싶다면?'이라는 가능성을 생각해보면 현재 상황에서는 대응이 어려울 것 같고,
- 훅은 가능하면 화면에 '렌더'하는 부분이 아닌 기능적인 부분을 분리해두는 역할로 보는 것이 컴포넌트와 구분짓기 좋지 않나 해서요     

제 개인 생각일 수 있으니 이후 프로젝트도 해보시면서 마르코의 기준을 정리해보시면 좋겠습니다! 🙂🙏

wonsss replied:
상세한 답변을 주셔서 감사합니다!! 저도 스낵바와 같은 UI 로직은 커스텀 훅으로 묶지 않는 편이 좋을 것 같다고 느꼈어요. 말씀해주신 내용 바탕으로 저만의 기준을 정리해보겠습니다! 🙏

woowapark commented on src/components/CartProductItem/useDeleteProductAPI.js:
```diff
@@ -0,0 +1,34 @@
+import useSnackbar from '../../hooks/useSnackbar';
+
+// DONE 5. delete 장바구니 내 선택된 상품 삭제
+const useDeleteProductAPI = productId => {
```
(요건 그냥 생각만!)    
API 호출부를 상세하게 잘 분리해주셨네요! 🙂
다만 보다보니 너무 HTTP method 하나하나 단위로 나뉜 게 아닐까 하는 생각이 들었는데요 🤔 method 단위보다는 어떤 대상을 다루는지에 대한 단위로 보자면      
`useProduct` 정도로 묶어볼 수도 있지 않을까요? (이전 버전에서 약간 다른 역할이지만 useProduct 훅이 있긴 했더라구요?ㅋㅋ)    
 product 상태에 관련된 곳에서는 공통으로 `useProduct`를 사용하고,    
 - add
 - delete 
 - increaseQuantity 
 - decreaseQuantity      
 
 와 같은 인터페이스를 제공해주면 어떨까요?      
 요건 그냥 의견으로 남겨보는 것이고, 수정 범위도 클 수 있는 일이라 그냥 생각해보기 정도로 봐주세용ㅎㅎ

wonsss replied:
오 좋은 것 같아요! 지금처럼 HTTP 메서드 단위보다는 대상 단위로 묶은 커스텀훅에 대상에 관련된 HTTP 메서드들을 인터페이스로 제공해주는 방법!! 다음에 이렇게 적용해볼게요! 감사합니다 😄

woowapark commented on src/components/CartProductItem/useDeleteProductAPI.js:
```diff
@@ -0,0 +1,34 @@
+  const deleteProduct = useCallback(async () => {
+    try {
+      await apiClient.delete('/cart', { data: { productIds: [productId] } });
+      dispatch(doDeleteProductFromCart({ productId: productId }));
```
(19-20라인같은 부분은 미들웨어를 적용한다면 그냥 비동기 액션을 dispatch 하는 걸로 한번에 묶을 수 있겠네요 🙂)

woowapark commented on src/components/CartProductItem/useDeleteProductAPI.js:
```diff
@@ -0,0 +1,34 @@
+    try {
+      await apiClient.delete('/cart', { data: { productIds: [productId] } });
+      dispatch(doDeleteProductFromCart({ productId: productId }));
+      renderSnackbar(MESSAGE.REMOVE_CART_SUCCESS, 'SUCCESS');
```
renderSnackbar를 사용하는 입장에서 `'SUCCESS'`나 `'FAILED'`같은 문자열을 넘겨주어야 한다는 것을 어떻게 알 수 있을까요?

wonsss replied:
오 좋은 포인트네요. 제 생각엔 타입스크립트를 적용하면 해당 스낵바 사용 시 해당 문자열이 추천되기 때문에 알 수 있을 것 같아요. 이렇게 수정해봤어요. 이외에 혹시 적절한 다른 방법이 있을까요?

[refactor: snackbar에 TS 적용하여 인자로 받을 문자열 좁히기](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/794788bf8a90250119d4b4de8d0043dc008c2b63)

woowapark replied:
오 TS로 적용하실 거라면 의미도 좀 더 드러나게 요런 식으로 조금 더 해볼 수도 있을 것 같아요 :)  
- 참고: https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums
```ts
// 요 상수를 import해서 제공하는 옵션들 중에서만 고를 수 있게 해볼 수 있을 것 같고요~
export const STATUS = {
  SUCCESS: 'success',
  FAIL: 'fail'
} as const;

type SnackbarStatus = typeof STATUS[keyof typeof STATUS];

(message: string, status: SnackbarStatus)
```

woowapark commented on src/hooks/useLogout.js:
```diff
@@ -0,0 +1,28 @@
+    const { code, message } = customError;
+
+    // 토큰 유효기간 만료, 유효하지 않은 토큰, 인증이 필요한 접근
+    if ([1002, 1003, 1004].includes(code)) {
```
- 요 숫자들이 어떤 에러를 의미하는 코드인지 알 수 있을까요?-?     
주석이 없어도 코드만으로 이해할 수 있게 만들 방법으로 고민해보면 좋을 것 같습니당~!    
- `[1002, 1003, 1004].includes(code)`가 합쳐서 어떤 조건인지를 변수 혹은 함수로 한번 묶어서 의미를 나타낼 수도 있을 것 같고요

wonsss replied:
주석 없이도 코드만으로 이해할 수 있도록 코딩 시 고민해야겠네요! 감사합니다. 😄 

[refactor: 주석 없이 코드만으로 이해할 수 있게 변수 추가](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/6ebe1f16fb736b38fdb01522c00d8ac49dd4b4fd)

woowapark commented on src/hooks/usePutCartAPI.js:
```diff
@@ -0,0 +1,51 @@
+      setIsPutCartLoading(true);
+
+      try {
+        const response = await apiClient.put(`/cart/products/${productId}`, {
```
요런 엔드포인트들도 상수화 해두는 편이 나중의 화를 방지할 수 있을지 몰라요ㅋㅋ     
만약 변경된다면 문자열들을 따로따로 찾아서 수정을 하나씩 해주어야 하는 불편함이 있기도 하고, 휴먼 에러가 발생하기도 쉽고요 🙂

woowapark commented on src/hooks/usePutCartAPI.js:
```diff
@@ -0,0 +1,51 @@
+        setError(customError || error);
+        logoutByError(customError);
+        renderSnackbar(customError.message || error.message, 'FAILED');
+        setIsPutCartLoading(true);
```
이미 putCart 내부이니 요런 식으로 -PutCart- 는 굳이 반복해서 안써도 괜찮을 것 같은데 어떠신가용? (커스텀 훅들에 비슷한 네이밍들이 많던데 동일하게!)      
만약 훅 이름을 조금 수정하게 되거나 하면 같이 맞춰서 바꿔주어야 할 것 같은데 그러기보다는 이미 훅 내부이니 `isLoading`정도여도 충분하지 않을까 싶습니당      
만약 이름이 겹치는 것 때문이라면 이 훅을 사용하는 쪽에서 꺼낼 때 이름을 재할당해주면 되고요

wonsss replied:
헉 맞아요 ㅎㅎㅎ 처음엔 저렇게 반복해서 적다가, 불필요한 것 같고 이름 중복은 사용부에서 이름 재할당하면되겠다 생각해서 단순히 `isLoading` 처럼 바꿨습니다. 그런데 처음에 작성해두고 안바뀐 부분들이 있네요. 이들 모두 isLoading 처럼 동일하게 수정했어요.

[refactor: loading 관련 네이밍을 동일하게 수정](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/58870934ffd2be950f967a57e89b2297d3efb1b7)

woowapark commented on src/hooks/useServerSelect.js:
```diff
@@ -0,0 +1,27 @@
+
+  const setServerApi = (name = '이프') => {
+    localStorage.setItem('api_name', name);
+    apiClient.defaults.baseURL = AUTH_API_URL[name];
```
직접 값에 할당하기보다는 apiClient에서 baseURL을 바꿀 수 있는 메서드를 가지고 있고 그걸 사용하도록 하는 게 어떨까요🙂?    
```suggestion
    apiClient.setBaseURL(AUTH_API_URL[name])
```    

apiClient의 정보를 요렇게 직접 접근해서 바꿔버리는 경우,    
나중에 어디에서 어떻게 변경하는 지를 추적하기 어려울 수 있을 것 같아요 🙂     
모듈로 묶었다면 해당 모듈의 정보는 모듈에서 제공하는 인터페이스를 통해서만 변경할 수 있게 해주는 게 어떨까 합니다~

wonsss replied:
앗! 고민하던건데! 이렇게 할 수 있겠네요. 
apiClient를 class로 만들어서 setBaseURL과 setAuth 메서드를 추가했어요.
[refactor: apiClient를 클래스로 만들어 setBaseURL과 setAuth 메서드 추가](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/dded94226e53883db3cc0bb00d6109cf4b736f1e)

woowapark commented on src/hooks/useTotalPrice.js:
```diff
@@ -0,0 +1,38 @@
+
+  useEffect(() => {
+    if (!isLoading && !isAuthenticated) {
+      renderSnackbar(MESSAGE.NO_AUTHORIZATION, 'FAILED');
```
인증 관련 처리는 요기에서 분리하는 편이 어떨까요?      
totalPrice를 계산하는 훅인데 너무 많은 책임을 지고 있는 게 아닐까 싶네요! 🙂

wonsss replied:
다시 살펴보니 해당 훅에서 인증 관련 처리 로직은 불필요하네요. 훅에서 불필요한 로직을 삭제했어요! 감사합니다. 😄 
[refactor: useTotalPrice 훅에서 불필요한 인증 관련 로직 삭제](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/2ea1a21e10c3c35934bde650ca9d529546a5bb49)

woowapark commented on src/hooks/useTotalPrice.js:
```diff
@@ -0,0 +1,38 @@
+      const product = shoppingCart.find(product => product.productId === id);
+      if (product) {
+        const { price, quantity } = product;
+        total += price * quantity;
```
요런 친구들은 [reduce](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce)를 적극 사용해볼까요~!

wonsss replied:
넵! 수정했습니다. [refactor: totalPrice 계산 로직을 reduce로 리팩토링](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/386a9d4e029fbbf6d9672381591de9f04d8ecd9c)

woowapark commented on src/hooks/useTotalPrice.js:
```diff
@@ -0,0 +1,38 @@
+  const navigate = useNavigate();
+  const [totalPrice, setTotalPrice] = useState(0);
+
+  const calculateTotalPrice = useCallback(() => {
```
- 지금 해주신 것처럼 커스텀 훅의 로직으로 계산할 수도 있고, 
- `useSelector`에 유사한 역할을 해줄 셀렉터 함수를 넘겨줄 수도 있고, 
- redux store에서 total을 store 내부 상태들 기준으로 계산하게 할 수도 있을 것 같은데요 🙂 (고러면 아마 shoppingCart에 변동이 있을 때마다 `dispatch(calculateTotalPrice())`같이 호출하게 되겠죠..?) 

요렇게 하기로 결정하신 이유가 있을까요?

wonsss replied:
커스텀 훅 분리에 집중하다보니,, ㅎㅎ 이런 코드가 나온 것 같아요. 말씀해주신 다른 대안들도 괜찮은 것 같습니다!

woowapark reviewed (changes requested):
마르코 안녕하세요~!         
대타로 찾아왔습니다 🙂 
 커스텀 훅 나눠주신 부분이나 여러모로 고민을 열심히 해주신 것 같다는 생각을 하면서 읽었습니당ㅋㅋ 마지막 미션까지 정말 고생 많으셨어요! 
 
### 질문 1에 대해서        

   저는 요 부분은 페어(아마 앞으로는 팀)와 컨벤션만 잘 맞추면 된다고 생각하는 편입니다 🙂         
 합의되었다는 전제 하에서는 여러 곳에서 사용할 게 아니라 한 군데에서만 쓰지만 로직을 분리하기 위한 용도로 만든 훅이라면 컴포넌트와 가까운 데 두는 것도 좋은 것 같아요.
질문 2,3은 코드 라인에 코멘트로 남겨두었어요!        

### 분리 기준     
- 요번엔 아주 잘게잘게 훅을 잘 분리해주신 것 같아요! 👏👏👏    
- 나누는 연습을 해보셨으니 이제 적절히 다시 합쳐본다면 어떻게 묶어볼 수 있을지 고민해보아도 좋을 것 같습니다 🙂      
  - 협업 미션을 하다보니 HTTP 메서드를 열심히 공부하면서ㅋㅋ 요렇게 분리하신 것이지 않았으려나~ 했는데, HTTP 메서드처럼 구체적인 구현사항, 동작 방식을 기준으로 나누기보다는 관심사 기준으로 나누는 걸 추천드리고 싶네요
- 더 나눌 부분은 나눠보고, 합칠 부분은 합쳐보고 하면서 마르코만의 기준을 앞으로 프로젝트 하면서도 계속 정리해나가보시면 좋을 것 같아요! 

### 기능 리뷰       
- 로딩되다가 타임아웃이 나거나, 네트워크가 끊기는 경우에 기본 데이터가 노출되는 것 같네용. 요런 에러도 빼먹지 않기!     
  - API가 실패하는 경우에도 코드상으로는 잘 처리해주셨는데요! 화면상에서는 에러가 발생했는지를 인지하기 어려울 것 같아 요 부분도 사용자 입장에서 한번 확인해보시면 좋을 것 같네요 🙂
- 로그인하지 않은 상태에서 장바구니 버튼을 클릭했을 때 동작이 아래와 같이 페이지에 따라 다른데요. 사실상 같은 케이스에 해당하니 일관된 동작으로 수정이 필요해보입니다. 
    - 상품 상세 > 장바구니 버튼 클릭 > 장바구니 페이지로 이동  + 토큰이 유효하지 않다는 토스트 메시지 노출
    - 상품 목록 > 각 상품의 장바구니 버튼 클릭 > 접근 권한이 없다는 토스트 메시지 노출 + 로그인 페이지로 이동
- 토스트 메시지의 경우에도, ‘토큰이 유효하지 않다’거나 ‘접근 권한이 없다’보다는 조금 더 사용자에 가까운 메시지를 고려해보면 어떨까요? 개발자가 아닌 일반 사용자 입장에서는 어떤 것때문에 지금 에러가 발생했다는 것인지 파악이 어려울 수 있을 것 같아요. ‘로그인이 필요하다’거나 꼭 요런 문구가 아니더라도 사용자 입장에서도 어떤 오류이고 + 어떤 걸 하면 되는 건지 이해할 수 있는 에러 메시지를 고민해보셔도 좋을 것 같아요 :) 실무에서는 개발자가 이걸 모두 정하진 않을 수 있지만, 그래도 같이 고민하고 다듬으면 좋으니까요~

기간이 워낙 촉박했어서 놓친 부분도 있을 것 같은데요 🙂      
리뷰는 어떤 부분들이 있었는지 정도 알아두고, 반영할 부분들만 선택적으로 반영해주시면 될 것 같습니다.       
그럼 해피 방학!

woowapark commented on src/components/ProductItem/index.jsx:
```diff
@@ -1,33 +1,36 @@
   const [isControllerOpen, setIsControllerOpen] = useState(false);
   const [clearTimer, setAutoCloseTimer, extendTimer] = useClose();
 
+  const { putCart } = usePutCartAPI();
```
아 요기도! `ProductItem` 같이 가장 트리 하위에 있는 개별 컴포넌트에서도 요런 식으로 api 호출, redux store에 대한 의존성이 있을 경우 api 변경이나 store에서의 데이터 구조 등이 변경될 경우 영향이 클 것 같은데요 🤔, 마르코 생각은 어떠신가용?

wonsss replied:
아 그렇군요. 이처럼 의존성이 클 경우 의존하는 부분에서 변경으로 인한 부작용이 있을 수 있겠네요. 흠.. 그런데 어떤 식으로 수정하면 좋을지 고민해도 감이 잘 잡히지 않네요ㅠㅠ

woowapark commented on src/utils/validator.ts:
```diff
@@ -1,23 +1,63 @@
+  password: string;
+};
+
 export const validateEmail = (email: string) => {
```
타이핑 시도해보신 거니 리턴 타입도 명시해두면 어떨까요 🙂

wonsss replied:
넵! 해당 파일의 타입스크립트 보완했습니다. 이번 방학에 TS 공부 열심히 하려구요 ㅎㅎ 😄 
[refactor: validator.ts 반환 타입 등 보완](https://github.com/woowacourse/react-shopping-cart-prod/pull/74/commits/5c0c064c895d32314fcbc755bf7b4b8b478164b1)

woowapark replied:
`never`타입에 대해서도 학습해보시면 좋을 것 같아요~! 🙂

wonsss committed:
chore: 불필요한 명령어 삭제

wonsss committed:
refactor: reducer 파일을 카멜케이스 네이밍 방식으로 변경

wonsss committed:
refactor: action creator 네이밍 컨벤션 변경('do' prefix 삭제)

wonsss committed:
refactor: App.js에서 useCallback 적용 및 오타 수정

wonsss committed:
refactor: usePutCartAPI커스텀 훅에 UI의 기능 옮기기

https://github.com/woowacourse/react-shopping-cart-prod/pull/74#discussion_r896451581

wonsss committed:
feat: PrivateRoute, PublicRoute 적용

wonsss committed:
refactor: snackbar에 TS 적용하여 인자로 받을 문자열 좁히기

https://github.com/woowacourse/react-shopping-cart-prod/pull/74#discussion_r896725589

wonsss committed:
refactor: 주석 없이 코드만으로 이해할 수 있게 변수 추가

https://github.com/woowacourse/react-shopping-cart-prod/pull/74#discussion_r896760250

wonsss committed:
refactor: loading 관련 네이밍을 동일하게 수정

https://github.com/woowacourse/react-shopping-cart-prod/pull/74#discussion_r896765242

wonsss committed:
refactor: apiClient를 클래스로 만들어 setBaseURL과 setAuth 메서드 추가

wonsss committed:
refactor: useTotalPrice 훅에서 불필요한 인증 관련 로직 삭제

https://github.com/woowacourse/react-shopping-cart-prod/pull/74#discussion_r896777711

wonsss committed:
refactor: totalPrice 계산 로직을 reduce로 리팩토링

wonsss committed:
refactor: validator.ts 반환 타입 등 보완

woowapark commented on src/apis/apiClient.js:
```diff
@@ -0,0 +1,27 @@
+const accessToken = getCookie('accessToken');
+const name = localStorage.getItem('api_name') || '이프';
+
+class apiClientClass {
```
`-Class` suffix가 꼭 필요할까요?

woowapark reviewed (approved):
마르코 마지막까지 수정하시느라 고생 많으셨어요! 🙂👍      
미션 마무리를 하고 마르코를 놓아드리기 위해 PR은 머지하겠습니다~!
추가로 남긴 코멘트들은 머지 후에 고민해보셔도 좋을 것 같고,    
머지 후에도 혹시 코멘트 관련해서 궁금하신 부분이나 추가 수정해보고 싶으신 부분이 있으시면     
코멘트에 마저 달아주시거나 DM으로 질문 주셔도 좋습니당~! 

이제 프로젝트에서 지금까지 잘 배워오신 것들 실전으로 재미나게 적용해보아요! 
그럼 이제 정말로 해피 방학! 🚀
